ls1intum/Artemis
# `General`: Notify users more reliably when there is a version update

### Checklist
#### General


#### Client

### Description

This PR was intially started by @krusche after he noticed that the service worker doesn't reliably update the client without a bunch of reloads. This PR adds an interval check to the service worker to look for updates, and adds a new button to the outdated alert that, if clicked, tells the service worker to activate the new version.

### Steps for Testing

Prerequisites:
- Two commits with a build; the 2nd commit should have a different version that the first one in both package.json and gradle.
You can use these prepared releases:
https://bamboo.ase.in.tum.de/deploy/viewDeploymentVersionCommits.action?versionId=486310120 ( b26f4c0 )
https://bamboo.ase.in.tum.de/deploy/viewDeploymentVersion.action?versionId=486310121 ( e0ba3cc )

Steps:
- Deploy the first commit to a test server. Wait for it to launch completely, then load the page.
- Make sure the service worker is installed (check dev tools)
- Verify that the correct version is displayed in the page top bar
- Keep the page open. Do not reload.
- Deploy the second commit to the server. Wait. Do not reload.
- The page will probably show an alert about it being disconnected. That's normal. Do not reload.
- Wait for the blue "your app is outdated" alert to appear. Do not reload, don't click the button yet.
- Navigate around in the app. Verify that the blue outdated alert keeps being displayed.
- Close the alert by clicking the small "X".
- Continue to navigate around. Verify that the alert reappears after some time.
- Finally, press the "Update" button in the alert.
- Check that the app updated to the new version (check version in the top bar).

### Review Progress





#### Code Review
#### Manual Tests
